assembly/index.ts

```diff
@@ -1,4 +1,221 @@
-import { JSONDecoder } from './decoder';
+import { JSONDecoder, JSONHandler } from './decoder';
 import { JSONEncoder } from './encoder';
 
+export namespace JSON {
+    export enum Val_Type {
+        STRING = 0,
+        NUMBER = 1,
+        BOOL = 2,
+        NULL = 3,
+        ARRAY = 4,
+        OBJECT = 5
+    }
+
+    export abstract class Value {
+        constructor(public type: Val_Type) { }
+
+        static String(str: string): Value {
+            return new Str(str)
+        }
+        static Number(num: i64): Value {
+            return new Number(num)
+        }
+        static Bool(b: bool): Value {
+            return new Bool(b)
+        }
+        static Null(): Value {
+            return new Null();
+        }
+        static Array(): Value {
+            return new Arr();
+        }
+        static Object(): Value {
+            return new Object();
+        }
+
+        get isString(): bool {
+            return this.type == Val_Type.STRING;
+        }
+
+        get isObject(): bool {
+            return this.type == Val_Type.OBJECT;
+        }
+
+        get isArray(): bool {
+            return this.type == Val_Type.ARRAY;
+        }
+
+        get isNumber(): bool {
+            return this.type == Val_Type.NUMBER;
+        }
+
+        get isBool(): bool {
+            return this.type == Val_Type.BOOL;
+        }
+
+        get isNull(): bool {
+            return this.type == Val_Type.NULL;
+        }
+    }
+
+    export class Str extends Value {
+        constructor(public str: string) {
+            super(Val_Type.STRING);
+        }
+    }
+
+    export class Number extends Value {
+        constructor(public num: i64) {
+            super(Val_Type.NUMBER);
+        }
+    }
+
+    export class Null extends Value {
+        constructor() {
+            super(Val_Type.NULL);
+        }
+    }
+
+    export class Bool extends Value {
+        constructor(public bool: bool) {
+            super(Val_Type.BOOL);
+        }
+    }
+
+    export class Arr extends Value {
+        array: Array<Value> = new Array<Value>();
+        constructor() {
+            super(Val_Type.ARRAY);
+        }
+    }
+
+    export class Object extends Value {
```

Comment:
s/Object/JSONObject/

Reply:
I think `JsonObject` is more readable.

Reply:
This would mean it would be `json.JSONObject`, which seems redundant.  Should I just remove the namespace?